### DISCUSSION TITLE
RFC: class based approach inspired by type-graphql

### DISCUSSION THREAD

**Discussion by ywarezk** [trpc/trpc#4492] [Ideas]
The following suggestion is heavily inspired by the [type-graphql](https://github.com/MichalLytek/type-graphql) library.  
The idea is to bring a similar approach to trpc and allow developers to develop their trpc api using class based approach.  
To help doing this a set of decorators will be supplied.

Here is a simple example of what I thought:

```typescript
import { Router,  Query, Auth, Input, Output} from '@trpc/decorators'; 
import { z } from 'zod';

@Router()
export class HelloRouter {

    private _helperMethod(
        @Ctx() context: Context
    ) { 
        ...
    }

    @Query()
    @Auth(['admin', 'user'])
    @Input(z.object({
        hello: 'world'
    }))
    @Output(z.object({
        foo: 'bar'
    }))
    async helloProcedure({ input, ctx }) {
        ...
    }
}
```

Few advantages with this approach:
- cleaner way to create base classes with shared functionality
- context can be injected to other helper methods in the class
- **type-graphql** are also allowing to embed [typedi](https://github.com/typestack/typedi) which allows us to easily embed dependency injection
- easier to reuse other router classes

**Comment by KATT** [comment]
Hey @ywarezk, I find this approach interesting and I've played around with similar ideas before, see #3727.

My main question is if the decorators in TypeScript are enough yet to support this, I'm not sure if e.g. `@Auth()` & `@Input` can automatically infer the type of `input` and `ctx` in the options object and if `@Output` can enforce the return type of the function? If so, I'd love to see a TS playground of a small POC where you use decorators as builders

**Reply by ywarezk** [reply]
working currently on a POC in this repository: [trpceez](https://github.com/ywarezk/trpceez)

I have my doubts as well, and I'm really skeptic if decorators can achieve this kind of checks on the actual class methods.
But I do think that if decorators cannot type check the method, then decorators + typescript plugin can probably do the job.

`tsconfig.json`


```json
{
    compilerOptions: {
        plugins: [{
            name: "trpceez"
        }]
    }
}
```

`my-router.router.ts`

```typescript

@Router()
class MyRouter {

    @Query()
    @Output(z.string())
    async myHelloQuery() {
        return 10; // the plugin should mark an error here
    }
}
```

**Comment by rvalledorjr** [comment]
+1

**Reply by KATT** [reply]
Open for PRs

**Comment by JacobFV** [comment]
any status?

**Comment by denz93** [comment]
@KATT @ywarezk 
I just create a [trpc-decorator](https://github.com/denz93/trpc-decorator) library to work with tRPC v10. I can create a PR if it is adopted. 

`npm install trpc-decorator`

**Comment by JacobFV** [comment]
rly cool

On Fri, Oct 25, 2024 at 6:09 PM Nhan Bach ***@***.***> wrote:

> @KATT <https://github.com/KATT> @ywarezk <https://github.com/ywarezk>
> I just create a trpc-decorator <https://github.com/denz93/trpc-decorator>
> library to work with tRPC v10. I can create a PR if it is adopted.
>
> npm install trpc-decorator
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/trpc/trpc/discussions/4492#discussioncomment-11057771>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJTZS2LQLFJ2MHSFO7LUV3TZ5LTVVAVCNFSM6AAAAAAY5WB4XWVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMBVG43TOMI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Jacob Valdez
